sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolvePartitionSpec.scala

```diff
@@ -65,31 +65,8 @@ object ResolvePartitionSpec extends Rule[LogicalPlan] {
       conf.resolver)
 
     val partValues = partSchema.map { part =>
-      val partValue = normalizedSpec.get(part.name).orNull
-      if (partValue == null) {
-        null
-      } else {
-        // TODO: Support other datatypes, such as DateType
-        part.dataType match {
-          case _: ByteType =>
-            partValue.toByte
-          case _: ShortType =>
-            partValue.toShort
-          case _: IntegerType =>
-            partValue.toInt
-          case _: LongType =>
-            partValue.toLong
-          case _: FloatType =>
-            partValue.toFloat
-          case _: DoubleType =>
-            partValue.toDouble
-          case _: StringType =>
-            partValue
-          case _ =>
-            throw new AnalysisException(
-              s"Type ${part.dataType.typeName} is not supported for partition.")
-        }
-      }
+      val raw = normalizedSpec.get(part.name).orNull
+      Cast(Literal.create(raw, StringType), part.dataType, Some(conf.sessionLocalTimeZone)).eval()
```

Comment:
Is this matched to the V1 partitioning type coercion? I remember it has a bit different rules, see `PartitioningUtils.inferPartitionColumnValue`.

Reply:
We don't have unified tests for V1 and V2 `ALTER TABLE .. ADD/DROP PARTITION` at the moment. I plan to do that soon. As soon as we have such tests we will see the differences and fix them.

For now, I just try to make implementation simpler - cast partition values according to the partition schema.

Reply:
I think V1 does the same, see https://github.com/apache/spark/blob/dfa6fb46f4238792bff6a0201da201be1b42620e/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala#L149-L165

Let me extract the code to `PartitioningUtils`, and re-use it in V2.

Reply:
Gotya

Reply:
I reused the code from DSv1 https://github.com/apache/spark/pull/30482 and fixed an issue.